// standardized-audio-context/build/es2018/factories/channel-splitter-node-constructor.js
import { getNativeContext } from '../helpers/get-native-context';
const DEFAULT_OPTIONS = {
    channelCount: 6,
    channelCountMode: 'explicit',
    channelInterpretation: 'discrete',
    numberOfOutputs: 6
};
const sanitizedOptions = (options) => {
    return { ...options, channelCount: options.numberOfOutputs };
};
export const createChannelSplitterNodeConstructor = (createChannelSplitterNodeRenderer, createNativeChannelSplitterNode, isNativeOfflineAudioContext, noneAudioDestinationNodeConstructor) => {
    return class ChannelSplitterNode extends noneAudioDestinationNodeConstructor {
        constructor(context, options = DEFAULT_OPTIONS) {
            const nativeContext = getNativeContext(context);
            const mergedOptions = sanitizedOptions({ ...DEFAULT_OPTIONS, ...options });
            const nativeChannelSplitterNode = createNativeChannelSplitterNode(nativeContext, mergedOptions);
            const channelSplitterNodeRenderer = ((isNativeOfflineAudioContext(nativeContext))
                ? createChannelSplitterNodeRenderer()
                : null);
            super(context, 'passive', nativeChannelSplitterNode, channelSplitterNodeRenderer);
        }
    };
};
